src/ray/core_worker/transport/direct_task_transport.h

```diff
@@ -193,36 +227,123 @@ class CoreWorkerDirectTaskSubmitter {
   // worker using a single lease.
   const uint32_t max_tasks_in_flight_per_worker_;
 
+  // work_stealing_ indicates whether the work stealing mode is enabled
+  const bool work_stealing_;
+
   /// A LeaseEntry struct is used to condense the metadata about a single executor:
   /// (1) The lease client through which the worker should be returned
   /// (2) The expiration time of a worker's lease.
   /// (3) The number of tasks that are currently in flight to the worker
-  /// (4) The resources assigned to the worker
-  /// (5) The SchedulingKey assigned to tasks that will be sent to the worker
+  /// (4) A boolean that indicates whether we have launched a StealTasks request, and we
+  /// are waiting for the stolen tasks (5) The resources assigned to the worker (6) The
+  /// SchedulingKey assigned to tasks that will be sent to the worker
   struct LeaseEntry {
     std::shared_ptr<WorkerLeaseInterface> lease_client;
     int64_t lease_expiration_time;
     uint32_t tasks_in_flight;
+    bool currently_stealing;
+    int64_t stolen_tasks_to_wait_for;
+    uint32_t steal_tasks_request_pending;
     google::protobuf::RepeatedPtrField<rpc::ResourceMapEntry> assigned_resources;
     SchedulingKey scheduling_key;
 
     LeaseEntry(
         std::shared_ptr<WorkerLeaseInterface> lease_client = nullptr,
         int64_t lease_expiration_time = 0, uint32_t tasks_in_flight = 0,
+        bool currently_stealing = false, int64_t stolen_tasks_to_wait_for = 0,
+        uint32_t steal_tasks_request_pending = 0,
         google::protobuf::RepeatedPtrField<rpc::ResourceMapEntry> assigned_resources =
             google::protobuf::RepeatedPtrField<rpc::ResourceMapEntry>(),
         SchedulingKey scheduling_key = std::make_tuple(0, std::vector<ObjectID>(),
                                                        ActorID::Nil()))
         : lease_client(lease_client),
           lease_expiration_time(lease_expiration_time),
           tasks_in_flight(tasks_in_flight),
+          currently_stealing(currently_stealing),
+          stolen_tasks_to_wait_for(stolen_tasks_to_wait_for),
+          steal_tasks_request_pending(steal_tasks_request_pending),
           assigned_resources(assigned_resources),
           scheduling_key(scheduling_key) {}
 
     // Check whether the pipeline to the worker associated with a LeaseEntry is full.
     bool PipelineToWorkerFull(uint32_t max_tasks_in_flight_per_worker) const {
       return tasks_in_flight == max_tasks_in_flight_per_worker;
     }
+
+    // Check whether the worker is a thief who is in the process of stealing tasks.
+    // Knowing whether a thief is currently stealing is important to prevent the thief
+    // from initiating another StealTasks request or from being returned to the raylet
+    // until stealing has completed.
+    bool WorkerIsStealing() const {
```

Comment:
Actually, could you inline these methods into the class? Since they're not doing much, I think it's actually more readable to be able to see what variables are getting touched when reading the main code.

Reply:
Fixed this in commit: https://github.com/ray-project/ray/pull/13570/commits/a114b4e17a1cdb773152477c18b24298e56c31aa